### PR TITLE
GE-915: X-Api-Key header authentication

### PR DESCRIPTION
**Pull request by ruands**

### CODE REVIEW

File: swagger_server/__main__.py [hunk 1]
```diff
@@ -6,18 +6,19 @@
...
 
 DB = SQLAlchemy()
 
+
 def main():
     app = connexion.App(__name__, specification_dir='./swagger/')
     app.app.json_encoder = encoder.JSONEncoder
     app.add_api('swagger.yaml', arguments={'title': 'User Data API'})
-    app.app.config = models.app.config
+    app.app.config = models.APP.config
```
Comment (AltusBarry) [inline swagger_server/__main__.py]: UDS does not have an all caps APP in models. ACS does.

Comment (AltusBarry) [reply swagger_server/__main__.py]: Tests do not cover __main__.py. Always check the apps startup manually.

File: swagger_server/test/__init__.py [hunk 2]
```diff
@@ -19,9 +23,10 @@ def create_app(self):
         #logging.getLogger('connexion.operation').setLevel('ERROR')
         app = connexion.App(__name__, specification_dir='../swagger/')
         app.app.json_encoder = JSONEncoder
-        app.app.config = models.app.config
+        app.app.config = models.APP.config
```
Comment (AltusBarry) [inline swagger_server/test/__init__.py]: Hmm now I'm confused. I know one app uses caps and the other not, has something changed there?

File: swagger_server/test/test_exception_handlers.py [hunk 3]
```diff
@@ -36,6 +26,8 @@ def setUp(self):
             action="create"
         )
 
+        self.headers = {API_KEY_HEADER: "test-api-key"}
```
Comment (AltusBarry) [inline swagger_server/test/test_exception_handlers.py]: Check about altering client in setUp

File: swagger_server/test/test_middleware.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            "site_id": random.randint(2, 2000000),
+            "schema": {"test": "data"}
+        }
+        self.sitedataschema_model = db_actions.crud(
+            model="SiteDataSchema",
+            api_model=SiteDataSchema,
+            data=self.sitedataschema_data,
+            action="create"
+        )
+
```
Comment (AltusBarry) [inline swagger_server/test/test_middleware.py]: Remove open line.

File: user_data_store/__init__.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import os
```
Comment (AltusBarry) [inline user_data_store/__init__.py]: Move UDS tests out to `user_data_store/tests/`

File: user_data_store/migrations/env.py [hunk 6]
```diff
@@ -13,6 +15,10 @@
 fileConfig(config.config_file_name)
 logger = logging.getLogger('alembic.env')
 
+orig_environ = dict(os.environ)
+orig_environ["ALLOWED_API_KEYS"] = "test-api-key"
```
Comment (AltusBarry) [inline user_data_store/migrations/env.py]: Why do migrations require the env value?

File: user_data_store/models.py [hunk 7]
```diff
@@ -1,14 +1,18 @@
...
 from sqlalchemy.dialects.postgresql import UUID
 from sqlalchemy.ext.compiler import compiles
 from sqlalchemy.sql import expression
 
-from user_data_store.settings import app
+from . import settings
 
-db = SQLAlchemy(app)
-migrate = Migrate(app, db)
+APP = Flask(__name__)
```
Comment (AltusBarry) [inline user_data_store/models.py]: Ah, I see. Ignore my comments regarding APP.

File: user_data_store/settings.py [hunk 8]
```diff
@@ -1,11 +1,8 @@
...
 DEFAULT_API_LIMIT = os.environ.get("DEFAULT_API_LIMIT", 50)
-
-app = Flask(__name__)
-app.config["SQLALCHEMY_DATABASE_URI"] = os.environ.get(
+DB_URI = os.environ.get(
         "DB_URI", "postgresql+psycopg2://user_data_store:password@127.0.0.1:5432/user_data_store")
-app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
+
+API_KEY_HEADER = "X-API-KEY"
+ALLOWED_API_KEYS = set(os.getenv("ALLOWED_API_KEYS").split(","))
```
Comment (AltusBarry) [inline user_data_store/settings.py]: Don't use `getenv` use `os.environ["ALLOWED_API_KEYS"]`

File: user_data_store/migrations/env.py (hunk 6 above)
Comment (ruands) [reply user_data_store/migrations/env.py]: Never got removed once I fixed the issue. Removing.

File: user_data_store/__init__.py (hunk 5 above)
Comment (ruands) [reply user_data_store/__init__.py]: Tests are moved out there already

File: swagger_server/test/test_exception_handlers.py (hunk 3 above)
Comment (ruands) [reply swagger_server/test/test_exception_handlers.py]: Looked into the werkzeug client and from what I can see this is the best way to do it. Else I have to set up an Environment builder for every call which will defeat the purpose of this change.

File: user_data_store/__init__.py (hunk 5 above)
Comment (AltusBarry) [reply user_data_store/__init__.py]: Then this should be in the tests/__init__.py

Comment (AltusBarry) [reply user_data_store/__init__.py]: Having it like this runs the risk of overriding the env each time uds is called by something.

**Comment by cobusc** [comment]
:+1:

**Pull request merged**